File: advent2024.kt/src/main/kotlin/solutions/Day02.kt
package net.libexec.adventofcode.solutions

import net.libexec.adventofcode.Day
import net.libexec.adventofcode.util.collections.isAscending
import net.libexec.adventofcode.util.collections.isDescending
import net.libexec.adventofcode.util.collections.splitIntegers
import kotlin.math.abs

class Day02(loadData: Boolean = true) : Day("02", loadData) {
    override fun solveA(): String {
        var safe = 0

        input.forEach {
            val values = it.splitIntegers(' ')
            var isSafe = true

            if (values.isAscending() || values.isDescending()) {
                for (i in 0..<values.size-1) {
                    val diff = abs(values[i] - values[i + 1])
                    if (diff != 1 && diff != 2 && diff != 3) {
                        isSafe = false
                    }
                }

                if (isSafe) safe++
            }
        }

        return safe.toString()
    }

    override fun solveB(): String {
        var safe = 0

        input.forEach {
            val values = it.splitIntegers(' ')

            if (values.isAscending() || values.isDescending()) {
                val candidates = getCandidatesB(values)
                var foundGood = false

                for (vi in candidates.size-1 downTo 0) {
                    val v = candidates[vi]
                    var failures = 0

                    for (i in 0..<v.size - 1) {
                        val diff = abs(v[i] - v[i + 1])
                        if (diff != 1 && diff != 2 && diff != 3) {
                            failures++
                        }
                    }

                    if (!foundGood && failures <= 2) {
                        foundGood = true
                        break
                    }
                }

                if (foundGood) safe++
            }
        }

        return safe.toString()
    }

    private fun getCandidatesB(v: List<Int>): List<List<Int>> {
        val candidates = ArrayList<List<Int>>()
        candidates.add(v)

        // Add copies with the first and
        candidates.add(v.slice(1..v.size-1))
        candidates.add(v.slice(0..v.size-2))

        for (i in 0..v.size-1) {
            candidates.add(v.slice(0..<i) + v.slice(i+1..v.size-1))
        }

        return candidates
    }
}

File: advent2024/advent2024/Day02.cs
using System.Diagnostics.CodeAnalysis;
using Libexec.Advent;
using Libexec.Advent.Extensions;
using Xunit.Abstractions;

namespace advent2024;

/// <summary>
/// 2024 day 2.
/// </summary>
/// <param name="output">A <see cref="ITestOutputHelper"/> to use for logging.</param>
/// <param name="isTest"><c>true</c> to load test data, <c>false</c> to load real data.</param>
/// <param name="fileSuffix">test file suffix.</param>
[SuppressMessage("ReSharper", "UnusedType.Global")]
public class Day02(ITestOutputHelper output, bool isTest = false, string fileSuffix = "") : Day(2, output, isTest, fileSuffix)
{
    /// <inheritdoc/>
    public override object PartA()
    {
        var safe = 0;

        foreach (var line in Input)
        {
            var values = line.ParseMany<int>().ToList();
            var isSafe = true;

            if (!values.IsAscending() && !values.IsDescending())
                continue;
            
            for (var i = 0; i < values.Count - 1; i++)
            {
                var diff = Math.Abs(values[i] - values[i + 1]);
                if (diff != 1 && diff != 2 && diff != 3)
                    isSafe = false;
            }

            if (isSafe)
                safe++;
        }

        return safe.ToString();
    }

    /// <inheritdoc/>
    public override object PartB()
    {
        var safe = 0;

        foreach (var line in Input)
        {
            var values = line.ParseMany<int>().ToList();

            var candidates = GetCandidates(values);
            
            foreach (var candidate in candidates)
            {
                var isSafe = true;
                var candidateArray = candidate.ToArray();
                
                if (!candidateArray.IsAscending() && !candidateArray.IsDescending())
                    continue;
                
                for (var i = 0; i < candidateArray.Length - 1; i++)
                {
                    var diff = Math.Abs(candidateArray[i] - candidateArray[i + 1]);
                    if (diff != 1 && diff != 2 && diff != 3)
                        isSafe = false;
                }

                if (!isSafe)
                    continue;

                safe++;
                break;
            }
        }

        return safe.ToString();
    }

    [SuppressMessage("Performance", "CA1859:Use concrete types when possible for improved performance")]
    private static IEnumerable<IEnumerable<int>> GetCandidates(IEnumerable<int> values)
    {
        var v = values.ToArray();
        var candidates = new List<IEnumerable<int>>
        {
            // The original list
            v,
            
            // First digit removed
            v[1..].ToList(),
            
            // Last digit removed
            v[..^1].ToList()
        };

        for (var i = 1; i < v.Length - 1; i++)
        {
            var newCandidate = new List<int>();
            newCandidate.AddRange(v[..i]);
            newCandidate.AddRange(v[(i+1)..]);
            
            candidates.Add(newCandidate);
        }

        return candidates;
    }
}